beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java

```diff
@@ -0,0 +1,196 @@
+
+      this.name = name;
+      this.position = position;
+      this.limit = limit;
+      this.repositoryFactory = repositoryFactory;
+      this.readerFactory = readerFactory;
+    }
+
+    @ProcessElement
+    public ProcessContinuation process(
+        RestrictionTracker<OffsetRange, Offset> tracker, OutputReceiver<StreamElement> output) {
+
+      try (ObserveHandle handle = startObserve(tracker.currentRestriction())) {
+        int i = 0;
+        while (!Thread.currentThread().isInterrupted()
+            && observer.getWatermark() < Watermarks.MAX_WATERMARK) {
+
+          StreamElement element = observer.take();
+          if (element == null) {
+            break;
+          }
+          boolean isFirst = i++ == 0;
+          Offset offset = Objects.requireNonNull(observer.getLastReadContext().getOffset());
+          if (externalizableOffsets
+              && !tracker.currentRestriction().isStartInclusive()
+              && isFirst) {
+            // skip the exclusive first offset
+            continue;
+          }
+          if (!tracker.tryClaim(offset)) {
+            return ProcessContinuation.stop();
+          }
+          output.outputWithTimestamp(element, Instant.ofEpochMilli(element.getStamp()));
+        }
+        observer.stop();
+      }
+      boolean terminated = observer.getWatermark() >= Watermarks.MAX_WATERMARK;
+      return terminated ? ProcessContinuation.stop() : ProcessContinuation.resume();
+    }
+
+    private ObserveHandle startObserve(OffsetRange restriction) {
+      CommitLogReader reader = readerFactory.apply(repositoryFactory.apply());
+      this.externalizableOffsets = reader.hasExternalizableOffsets();
+      observer = BlockingQueueLogObserver.create(name, limit, Watermarks.MIN_WATERMARK);
+      return reader.observeBulkOffsets(
+          Collections.singletonList(restriction.getStartOffset()), observer);
+    }
+
+    @GetInitialRestriction
+    public OffsetRange initialRestriction() {
+      return OffsetRange.initialRestriction();
+    }
+
+    @SplitRestriction
+    public void splitRestriction(
+        @Restriction OffsetRange restriction, OutputReceiver<OffsetRange> splits) {
+
+      if (restriction.isInitial()) {
+        CommitLogReader reader = readerFactory.apply(repositoryFactory.apply());
+        // compute starting offsets from commit log reader
+        List<Partition> partitions = reader.getPartitions();
+        try (ObserveHandle handle =
+            reader.observeBulkPartitions(partitions, position, droppingObserver())) {
+          ExceptionUtils.ignoringInterrupted(handle::waitUntilReady);
+          handle.getCurrentOffsets().forEach(o -> splits.output(OffsetRange.startingFrom(o)));
+        }
+      } else {
+        splits.output(restriction);
+      }
+    }
+
+    @GetRestrictionCoder
+    public Coder<OffsetRange> getRestrictionCoder() {
+      return SerializableCoder.of(OffsetRange.class);
+    }
+
+    private LogObserver droppingObserver() {
```

Comment:
can you please explain the whole "construct" with dropping observer and fetching offsets in javadoc? it's a bit tricky to understand